# House Prices - Advanced Regression Techniques
This is my second submission in Kaggle so if you find any mistakes, corrections, or suggestions, feel free to mention them.

## Initial Setup

### Import Statements

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Import the Dataset

In [2]:
train_data = pd.read_csv("./datasets/train.csv")
test_data = pd.read_csv("./datasets/test.csv")

### Stylistic

In [3]:
%matplotlib inline
pd.set_option("display.max_rows", 500)
pd.set_option("display.max_columns", 500)
pd.set_option("display.width", 500)
plt.style.use("dark_background")
np.set_printoptions(precision=2)

### Files

In [4]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('./datasets/'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

./datasets/test.csv
./datasets/data_description.txt
./datasets/train.csv
./datasets/sample_submission.csv


### Exploring the Dataset

#### Description of Each Column

In [5]:
train_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2003          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA      PConc       Gd       TA           No          GLQ         706          Unf           0        150          856    GasA        Ex          Y      SBrkr       856       854             0   
1   2          20       RL         80.0     9600   Pave   NaN      Reg         Lvl    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam     1Story            6            8       1976          1976     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA        TA     CBlock       Gd       TA           Gd          ALQ         978          Unf           0        284         1262    GasA        Ex          Y      SBrkr      1262         0             0   
2   3          60       RL         68.0    11250   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2001          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA      PConc       Gd       TA           Mn          GLQ         486          Unf           0        434          920    GasA        Ex          Y      SBrkr       920       866             0   
3   4          70       RL         60.0     9550   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam     2Story            7            5       1915          1970     Gable  CompShg     Wd Sdng     Wd Shng       None         0.0        TA        TA     BrkTil       TA       Gd           No          ALQ         216          Unf           0        540          756    GasA        Gd          Y      SBrkr       961       756             0   
4   5          60       RL         84.0    14260   Pave   NaN      IR1         Lvl    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam     2Story            8            5       2000          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd        TA      PConc       Gd       TA           Av          GLQ         655          Unf           0        490         1145    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice  
0       1710             1             0         2         1             3             1          Gd             8        Typ           0         NaN     Attchd       2003.0          RFn           2         548         TA         TA          Y           0           61              0          0            0         0    NaN   NaN         NaN        0       2    2008       WD        Normal     208500  
1       1262             0             1         2         0             3             1          TA             6        Typ           1          TA     Attchd       1976.0          RFn           2        

In [6]:
test_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl        NAmes      Feedr       Norm     1Fam     1Story            5            6       1961          1961     Gable  CompShg     VinylSd     VinylSd       None         0.0        TA        TA     CBlock       TA       TA           No          Rec       468.0          LwQ       144.0      270.0        882.0    GasA        TA          Y      SBrkr       896         0             0   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam     1Story            6            6       1958          1958       Hip  CompShg     Wd Sdng     Wd Sdng    BrkFace       108.0        TA        TA     CBlock       TA       TA           No          ALQ       923.0          Unf         0.0      406.0       1329.0    GasA        TA          Y      SBrkr      1329         0             0   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl      Gilbert       Norm       Norm     1Fam     2Story            5            5       1997          1998     Gable  CompShg     VinylSd     VinylSd       None         0.0        TA        TA      PConc       Gd       TA           No          GLQ       791.0          Unf         0.0      137.0        928.0    GasA        Gd          Y      SBrkr       928       701             0   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl      Gilbert       Norm       Norm     1Fam     2Story            6            6       1998          1998     Gable  CompShg     VinylSd     VinylSd    BrkFace        20.0        TA        TA      PConc       TA       TA           No          GLQ       602.0          Unf         0.0      324.0        926.0    GasA        Ex          Y      SBrkr       926       678             0   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1         HLS    AllPub    Inside       Gtl      StoneBr       Norm       Norm   TwnhsE     1Story            8            5       1992          1992     Gable  CompShg     HdBoard     HdBoard       None         0.0        Gd        TA      PConc       Gd       TA           No          ALQ       263.0          Unf         0.0     1017.0       1280.0    GasA        Ex          Y      SBrkr      1280         0             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
0        896           0.0           0.0         1         0             2             1          TA             5        Typ           0         NaN     Attchd       1961.0          Unf         1.0       730.0         TA         TA          Y         140            0              0          0          120         0    NaN  MnPrv         NaN        0       6    2010       WD        Normal  
1       1329           0.0           0.0         1         1             3             1          Gd             6        Typ           0         NaN     Attchd       1958.0          Unf         1.0       312.0      

In [7]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

Some insights for this data are:
1. We have 1460 entries.
2. We have 81 Columns.
3. The Indexing for this column are from 0 to 1459.
4. Alley, FireplaceQu, PoolQC, Fence, MiscFeature are some column with sparse data (Most data are Zero).
5. Most of our data are string, which needed to be encoded.

## Data Cleaning

### Removing Null Values

In [8]:
num_nulls1 = train_data.isnull().sum()
percentage1 = (train_data.isnull().sum()/train_data.isnull().count())
missing_values1 = pd.concat([num_nulls1, percentage1], axis=1, keys=["Train_Total", "Train_Pecentage"]).sort_values(by="Train_Total", ascending=False)

num_nulls2 = test_data.isnull().sum()
percentage2 = (test_data.isnull().sum()/test_data.isnull().count())
missing_values2 = pd.concat([num_nulls2, percentage2], axis=1, keys=["Test_Total", "Test_Pecentage"]).sort_values(by="Test_Total", ascending=False)

total_null = pd.concat([missing_values1, missing_values2], axis=1, keys=["Train_Total", "Train_Pecentage", "Test_Total", "Test_Pecentage"])
total_null

Train_Total                 Train_Pecentage               
              Train_Total Train_Pecentage      Test_Total Test_Pecentage
PoolQC               1453        0.995205          1456.0       0.997944
MiscFeature          1406        0.963014          1408.0       0.965045
Alley                1369        0.937671          1352.0       0.926662
Fence                1179        0.807534          1169.0       0.801234
FireplaceQu           690        0.472603           730.0       0.500343
LotFrontage           259        0.177397           227.0       0.155586
GarageYrBlt            81        0.055479            78.0       0.053461
GarageCond             81        0.055479            78.0       0.053461
GarageType             81        0.055479            76.0       0.052090
GarageFinish           81        0.055479            78.0       0.053461
GarageQual             81        0.055479            78.0       0.053461
BsmtFinType2           38        0.026027            42.0       0.028787
BsmtExposure           38        0.026027            44.0       0.030158
BsmtQual               37        0.025342            44.0       0.030158
BsmtCond               37        0.025342            45.0       0.030843
BsmtFinType1           37        0.025342            42.0       0.028787
MasVnrArea              8        0.005479            15.0       0.010281
MasVnrType              8        0.005479            16.0       0.010966
Electrical              1        0.000685             0.0       0.000000
Id                      0        0.000000             0.0       0.000000
Functional              0        0.000000             2.0       0.001371
Fireplaces              0        0.000000             0.0       0.000000
KitchenQual             0        0.000000             1.0       0.000685
KitchenAbvGr            0        0.000000             0.0       0.000000
BedroomAbvGr            0        0.000000             0.0       0.000000
HalfBath                0        0.000000             0.0       0.000000
FullBath                0        0.000000             0.0       0.000000
BsmtHalfBath            0        0.000000             2.0       0.001371
TotRmsAbvGrd            0        0.000000             0.0       0.000000
GarageCars              0        0.000000             1.0       0.000685
GrLivArea               0        0.000000             0.0       0.000000
GarageArea              0        0.000000             1.0       0.000685
PavedDrive              0        0.000000             0.0       0.000000
WoodDeckSF              0        0.000000             0.0       0.000000
OpenPorchSF             0        0.000000             0.0       0.000000
EnclosedPorch           0        0.000000             0.0       0.000000
3SsnPorch               0        0.000000             0.0       0.000000
ScreenPorch             0        0.000000             0.0       0.000000
PoolArea                0        0.000000             0.0       0.000000
MiscVal                 0        0.000000             0.0       0.000000
MoSold                  0        0.000000             0.0       0.000000
YrSold                  0        0.000000             0.0       0.000000
SaleType                0        0.000000             1.0       0.000685
SaleCondition           0        0.000000             0.0       0.000000
BsmtFullBath            0        0.000000             2.0       0.001371
HeatingQC               0        0.000000             0.0       0.000000
LowQualFinSF            0        0.000000             0.0       0.000000
LandSlope               0        0.000000             0.0       0.000000
OverallQual             0        0.000000             0.0       0.000000
HouseStyle              0        0.000000             0.0       0.000000
BldgType                0        0.000000             0.0       0.000000
Condition2              0        0.000000             0.0       0.000000
Condition1              0        0.000000             0.0       0.000000

Let's remove columns that has 1 or more nulls in their data.

In [9]:
train_data.drop((missing_values1[missing_values1["Train_Total"] >=  1]).index, axis=1, inplace=True)

In [10]:
train_data.isnull().sum().max()

0

In [11]:
test_data.drop((missing_values2[missing_values2["Test_Total"] >=  1]).index, axis=1, inplace=True)
test_data.drop("Electrical", axis=1, inplace=True)

In [12]:
cols = test_data.columns.values
cols = np.append(cols, "SalePrice")
train_data = train_data[cols]

In [13]:
train_data["SalePrice"]

0       208500
1       181500
2       223500
3       140000
4       250000
         ...  
1455    175000
1456    210000
1457    266500
1458    142125
1459    147500
Name: SalePrice, Length: 1460, dtype: int64

In [14]:
test_data.shape

(1459, 46)

In [15]:
train_data.shape

(1460, 47)

Some columns that has Ordinal Values
1. PoolQC (Pool Quality)
2. Fence (Fence quality)
3. FireplaceQu (Fireplace quality)
4. LotFrontage (Linear feet of street)
5. GarageYrBlt (Garage Year Built)
6. GarageQual (Garage Quality)
7. GarageFinish
8. GarageCond (Garage Condition)
9. BsmtCond (Basement Condition)
10. BsmtQual (Basement Quality)
11. BsmtExposure (Basement Exposure)
12. BsmtFinType2 (Rating of Basement Finished)
13. BsmtFinType1 (Type 1)

In [16]:
test_data.isnull().sum().max()

0

### Removing Outliers

#### Get the Numeric and Category Columns

In [17]:
numeric_col_train = [col for col in train_data.columns if train_data[col].dtype == "int64" or train_data[col].dtype == "float64"][1:]
category_cols_train = [col for col in train_data.columns if train_data[col].dtype == "object"]

numeric_col_test = [col for col in test_data.columns if test_data[col].dtype == "int64" or train_data[col].dtype == "float64"][1:]
category_cols_test = [col for col in test_data.columns if test_data[col].dtype == "object"]

#### Train Data

In [18]:
before_removal = train_data.shape[0]

for feature in numeric_col_train:
    min_ = train_data[feature].mean() - (3 * train_data[feature].std())
    max_ = train_data[feature].mean() + (3 * train_data[feature].std())
    
    train_data = train_data[train_data[feature] >= min_]
    train_data = train_data[train_data[feature] <= max_]
    
print("Number of Observation Removed: {}".format(before_removal - train_data.shape[0]))

Number of Observation Removed: 358


## Data Preprocessing

### LabelEncoder()

#### Train Data

In [19]:
from sklearn.preprocessing import LabelEncoder

enc = LabelEncoder()
for feature in category_cols_train:
    train_data[feature] = enc.fit_transform(train_data[feature])

#### Test Data

In [20]:
from sklearn.preprocessing import LabelEncoder

enc = LabelEncoder()
for feature in category_cols_test:
    test_data[feature] = enc.fit_transform(test_data[feature])

### Standard Scaler

In [21]:
# from sklearn.preprocessing import StandardScaler
# ss_train = StandardScaler()
# ss_test = StandardScaler()
# # Train Data
# train_data[numeric_col_train] = ss_train.fit_transform(train_data[numeric_col_train])
# # Test Data
# test_data[numeric_col_test] = ss_test.fit_transform(test_data[numeric_col_test])

In [22]:
train_data.head()

Id  MSSubClass  LotArea  Street  LotShape  LandContour  LotConfig  LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  ExterQual  ExterCond  Foundation  Heating  HeatingQC  CentralAir  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  \
0   1          60     8450       1         3            3          4          0             5           2           1         0           4            7            5       2003          2003          1         0          2          2           2        1          0           1       856       854             0       1710         2         1             3             1             8           0           2           0           61              0          0            0         0        0   
1   2          20     9600       1         3            3          2          0            24           1           1         0           2            6            8       1976          1976          1         0          3          2           1        1          0           1      1262         0             0       1262         2         0             3             1             6           1           2         298            0              0          0            0         0        0   
2   3          60    11250       1         0            3          4          0             5           2           1         0           4            7            5       2001          2002          1         0          2          2           2        1          0           1       920       866             0       1786         2         1             3             1             6           1           2           0           42              0          0            0         0        0   
4   5          60    14260       1         0            3          2          0            15           2           1         0           4            8            5       2000          2000          1         0          2          2           2        1          0           1      1145      1053             0       2198         2         1             4             1             9           1           2         192           84              0          0            0         0        0   
6   7          20    10084       1         3            3          4          0            21           2           1         0           2            8            5       2004          2005          1         0          2          2           2        1          0           1      1694         0             0       1694         2         0             3             1             7           1           2         255           57              0          0            0         0        0   

   MoSold  YrSold  SaleCondition  SalePrice  
0       2    2008              4     208500  
1       5    2007              4     181500  
2       9    2008              4     223500  
4      12    2008              4     250000  
6       8    2007              4     307000

In [23]:
test_data.head()

Id  MSSubClass  LotArea  Street  LotShape  LandContour  LotConfig  LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  ExterQual  ExterCond  Foundation  Heating  HeatingQC  CentralAir  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  \
0  1461          20    11622       1         3            3          4          0            12           1           2         0           2            5            6       1961          1961          1         0          3          4           1        0          4           1       896         0             0        896         1         0             2             1             5           0           2         140            0              0          0          120         0        0   
1  1462          20    14267       1         0            3          0          0            12           2           2         0           2            6            6       1958          1958          3         0          3          4           1        0          4           1      1329         0             0       1329         1         1             3             1             6           0           2         393           36              0          0            0         0    12500   
2  1463          60    13830       1         0            3          4          0             8           2           2         0           4            5            5       1997          1998          1         0          3          4           2        0          2           1       928       701             0       1629         2         1             3             1             6           1           2         212           34              0          0            0         0        0   
3  1464          60     9978       1         0            3          4          0             8           2           2         0           4            6            6       1998          1998          1         0          3          4           2        0          0           1       926       678             0       1604         2         1             3             1             7           1           2         360           36              0          0            0         0        0   
4  1465         120     5005       1         0            1          4          0            22           2           2         4           2            8            5       1992          1992          1         0          2          4           2        0          0           1      1280         0             0       1280         2         0             2             1             5           0           2           0           82              0          0          144         0        0   

   MoSold  YrSold  SaleCondition  
0       6    2010              4  
1       6    2010              4  
2       3    2010              4  
3       6    2010              4  
4       1    2010              4

## Machine Learning Model

### Splitting the Dataset

In [24]:
X = train_data.drop(["Id", "SalePrice"], axis=1).values
y = train_data["SalePrice"].values

In [25]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y)

In [26]:
y_train

array([204000, 222000, 284000, 118000, 105000, 170000, 212000, 235000,
       146000,  94750, 274900, 112000, 130000,  75000, 149000, 144000,
       318000, 137500, 337500, 106000, 200000, 123500, 212000, 190000,
       120500, 148000, 133000, 151000, 128500, 374000,  81000, 171000,
       348000, 178900, 244000, 182900, 237500, 127500, 146800, 173000,
       127000,  64500, 138000, 135000, 174000, 336000, 160200, 163000,
       118000,  52500, 191000, 104900, 140000, 114500, 178000, 139000,
       185850, 139000, 196000, 160000, 159000, 225000, 183000, 143000,
       123000, 274725, 113000, 134432, 115000, 128200, 203000, 285000,
        97500, 143900, 183500,  88000, 116900, 318000, 111000, 202665,
       158500, 143000, 227000, 180000, 178000, 243000, 306000, 176000,
       175500, 187000, 185000, 172500, 153000, 139000, 229456, 258000,
        99500, 188700, 127500, 319900, 213000, 119000, 129000, 148500,
       194000, 325300, 277500, 152000, 139600, 142000, 174900,  93000,
      

### Fitting the Training Data to Model

#### Random Forest

In [27]:
from sklearn.ensemble import RandomForestRegressor

forest = RandomForestRegressor().fit(X_train, y_train)
acc_forest = forest.score(X_train, y_train)*100
acc_test_forest = forest.score(X_test, y_test)*100

In [28]:
from sklearn.ensemble import GradientBoostingRegressor

gradient = GradientBoostingRegressor().fit(X_train, y_train)
acc_gradient = gradient.score(X_train, y_train)*100
acc_test_gradient = gradient.score(X_test, y_test)*100

In [41]:
from sklearn.svm import SVR

svr = SVR(kernel = "rbf").fit(X_train, y_train)
acc_svr = svr.score(X_train, y_train)*100
acc_test_svr = svr.score(X_test, y_test)*100

In [30]:
from sklearn.neighbors import KNeighborsRegressor

knn = KNeighborsRegressor().fit(X_train, y_train)
acc_knn = knn.score(X_train, y_train)*100
acc_test_knn = knn.score(X_test, y_test)*100

In [31]:
from sklearn.linear_model import LinearRegression

linearReg = LinearRegression().fit(X_train, y_train)
acc_linearReg = linearReg.score(X_train, y_train)*100
acc_test_linearReg = linearReg.score(X_test, y_test)*100

In [32]:
from sklearn.svm import LinearSVR

linearSVR = LinearSVR().fit(X_train, y_train)
acc_linearSVR = linearSVR.score(X_train, y_train)*100
acc_test_linearSVR = linearSVR.score(X_test, y_test)*100

/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


#### Decision Tree

In [33]:
from sklearn.tree import DecisionTreeRegressor

tree = DecisionTreeRegressor(max_depth=5, random_state=0).fit(X_train, y_train)
acc_tree = tree.score(X_train, y_train)*100
acc_test_tree = tree.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_tree))
print("Accuracy on Test Set: {:.2f}".format(acc_test_tree))

Accuracy on Training Set: 86.57
Accuracy on Test Set: 73.12


#### Ridge Regression

In [34]:
from sklearn.linear_model import Ridge

ridge = Ridge(alpha=1).fit(X_train, y_train)
acc_ridge = tree.score(X_train, y_train)*100
acc_test_ridge = tree.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_ridge))
print("Accuracy on Test Set: {:.2f}".format(acc_test_ridge))

Accuracy on Training Set: 86.57
Accuracy on Test Set: 73.12


In [35]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

poly_reg = PolynomialFeatures(interaction_only=True)
X_poly = poly_reg.fit_transform(X_train)
reg = LinearRegression()
_ = reg.fit(X_poly, y_train)

acc_poly = reg.score(X_poly, y_train)*100
# acc_test_poly = reg.score(X_train, poly_reg.transform(X_test))*100

print("Accuracy on Training Set: {:.2f}".format(acc_poly))
# print("Accuracy on Training Set: {:.2f}".format(acc_test_poly))

Accuracy on Training Set: 98.19


### Predictions

#### X_test Predictions

In [36]:
from sklearn.metrics import r2_score

svr_pred = svr.predict(X_test)
svr_r2 = r2_score(y_test,svr_pred)*100

knn_pred = knn.predict(X_test)
knn_r2 = r2_score(y_test,knn_pred)*100

linearReg_pred = linearReg.predict(X_test)
linearReg_r2 = r2_score(y_test,linearReg_pred)*100

forest_pred = forest.predict(X_test)
forest_r2 = r2_score(y_test,forest_pred)*100

linearSVR_pred = linearSVR.predict(X_test)
linearSVR_r2 = r2_score(y_test,linearSVR_pred)*100

tree_pred = tree.predict(X_test)
tree_r2 = r2_score(y_test,tree_pred)*100

ridge_pred = ridge.predict(X_test)
ridge_r2 = r2_score(y_test,ridge_pred)*100

gradient_pred = gradient.predict(X_test)
gradient_r2 = r2_score(y_test,gradient_pred)*100

In [47]:
models = pd.DataFrame({
    'Model': ['KNN', 'Linear Regression', 
              'Random Forest', 'Linear SVC', 'Decision Tree', "Ridge Regression", "Gradient Boosting Regressor"],
    'Train Score': [acc_knn, acc_linearReg, acc_forest, acc_linearSVR, acc_tree, acc_ridge, acc_gradient],
    'Test Score': [acc_test_knn, acc_test_linearReg, acc_test_forest, acc_test_linearSVR, acc_test_tree, acc_test_ridge, acc_test_gradient]
#     'R2 Score': [svr_r2, knn_r2, linearReg_r2, forest_r2, linearSVR_r2, tree_r2, ridge_r2, gradient_r2]

                     })
sorted_models = models.sort_values(by=["Train Score", "Test Score"], ascending=False)

In [48]:
sorted_models

Model  Train Score  Test Score
2                Random Forest    98.150309   85.799807
6  Gradient Boosting Regressor    95.577839   86.349663
1            Linear Regression    87.002984   85.332850
4                Decision Tree    86.570809   73.124263
5             Ridge Regression    86.570809   73.124263
3                   Linear SVC    79.071732   78.517153
0                          KNN    77.678813   57.385262

#### Submission Predictions

In [39]:
predictions = gradient.predict(test_data.drop("Id", axis=1))

In [40]:
output = pd.DataFrame({'Id': test_data["Id"], 'SalePrice': predictions})
output.to_csv('my_submission.csv', index=False)
print("Your submission was successfully saved!")

Your submission was successfully saved!
